synapse/handlers/devicemessage.py

```diff
@@ -293,3 +296,65 @@ async def send_device_message(
                 # Enqueue a new federation transaction to send the new
                 # device messages to each remote destination.
                 self.federation_sender.send_device_messages(destination)
+
+    async def get_events_for_dehydrated_device(
+        self,
+        requester: Requester,
+        device_id: str,
+        since_token: Optional[str],
+        limit: int,
+    ) -> JsonDict:
+        """Fetches up to `limit` events sent to `device_id` starting from `since_token` and returns the new since token."""
+
+        user_id = requester.user.to_string()
+
+        # TODO(Nico): Figure out who should be allowed to use that endpoint.
+        # For now we just allow it for yourself and for the dehydrated device.
+        if device_id != requester.device_id:
+            dehydrated_device = await self.device_handler.get_dehydrated_device(user_id)
+            if dehydrated_device is not None and device_id != dehydrated_device[0]:
+                raise SynapseError(
+                    403,
+                    "Can only fetch messages for own device or dehydrated devices",
+                    Codes.UNAUTHORIZED,
+                )
+
+        since_stream_id = 0
+        if since_token and len(since_token) > 1 and since_token[0] == "d":
```

Comment:
we seem to be mixing multiple conditions together here:

* whether a since token was provided
* whether it meets the expected format

It sounds suspicious to me that we'll just ignore invalid since tokens; should we complain with an error?

Reply:
Well, the MSC doesn't really describe the behaviour there. What happens in that case on other endpoints? I guess if they return an error, I can just do that here too.